source/event_stream.c

```diff
@@ -0,0 +1,1118 @@
+            if (!encoded_buffer) {
+                return aws_raise_error(AWS_ERROR_OOM);
+            }
+
+            if (header->header_value_type == AWS_EVENT_STREAM_HEADER_UUID) {
+                aws_base64_encode((uint8_t *) header->header_value.static_val, header->header_value_len, encoded_buffer,
+                                  buffer_len);
+            } else {
+                aws_base64_encode((uint8_t *) header->header_value.variable_len_val, header->header_value_len,
+                                  encoded_buffer, buffer_len);
+            }
+            fprintf(fd, "      " DEBUG_STR_HEADER_VALUE "\"%s\"\n", encoded_buffer);
+            aws_mem_release(message->alloc, encoded_buffer);
+        }
+
+        fprintf(fd, "    }");
+
+        if (count < headers_count - 1) {
+            fprintf(fd, ",");
+        }
+        fprintf(fd, "\n");
+
+        count++;
+    }
+    aws_event_stream_headers_list_cleanup(&headers);
+    fprintf(fd, "  ],\n");
+
+    size_t payload_len = aws_event_stream_message_payload_len(message);
+    const uint8_t *payload = aws_event_stream_message_payload(message);
+    size_t encoded_len = 0;
+    aws_base64_compute_encoded_len(payload_len, &encoded_len);
+    char *encoded_payload = (char *) aws_mem_acquire(message->alloc, encoded_len);
+
+    if (!encoded_payload) {
+        return aws_raise_error(AWS_ERROR_OOM);
+    }
+
+    aws_base64_encode(payload, payload_len, encoded_payload, encoded_len);
+    fprintf(fd, "  \"payload\": \"%s\",\n", encoded_payload);
+    fprintf(fd, "  " DEBUG_STR_MESSAGE_CRC "%d\n}\n",
+            aws_event_stream_message_message_crc(message));
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_event_stream_headers_list_init(struct aws_array_list *headers, struct aws_allocator *allocator) {
+    assert(headers);
+    assert(allocator);
+
+    return aws_array_list_init_dynamic(headers, allocator, 4, sizeof(struct aws_event_stream_header_value_pair));
+}
+
+void aws_event_stream_headers_list_cleanup(struct aws_array_list *headers) {
+    assert(headers);
+
+    for (size_t i = 0; i < aws_array_list_length(headers); ++i) {
+        struct aws_event_stream_header_value_pair *header;
+        aws_array_list_get_at_ptr(headers, (void **) &header, i);
+
+        if (header->value_owned) {
+            aws_mem_release(headers->alloc, (void *) header->header_value.variable_len_val);
+        }
+    }
+
+    aws_array_list_clean_up(headers);
+}
+
+int aws_event_stream_add_string_header(struct aws_array_list *headers, const char *name, uint8_t name_len,
+                                       const char *value, uint16_t value_len, int8_t copy) {
+    struct aws_event_stream_header_value_pair header = {
+            .header_name_len = name_len,
+            .header_value_len = value_len,
+            .value_owned = copy,
+            .header_value_type = AWS_EVENT_STREAM_HEADER_STRING
+    };
+
+    memcpy((void *) header.header_name, (void *) name, (size_t) name_len);
```

Comment:
should write this as null terminated string.

Reply:
this is addressed by the new byte_buf api.